### PR TITLE
fix: Minify for partial DFAs

### LINKED ISSUE #182: Minimization of DFA from finite language makes the language infinite
While using the `DFA.from_finite_language()` method to create a basic DFA from a finite language and then applying dfa.minify(), the resulting DFA incorrectly introduces a loop that allows an infinite number of accepted words. This behavior deviates from the expected result where the minimized DFA should recognize the same finite language as the unminimized DFA.

**Steps to Reproduce:**

1. Create a DFA from a finite language using DFA.from_finite_language() with the language {'ab', 'abcb'} and input symbols {'a', 'b', 'c'}.
1. Print the maximum word length and the DFA representation.
1. Apply the minify() method on the DFA.
1. Print the maximum word length and the minimized DFA representation.

**Code**

```python
dfa = DFA.from_finite_language(
    language={'ab', 'abcb'},
    input_symbols={'a', 'b', 'c'}
)
print("unminified maximum_word_length:", dfa.maximum_word_length())
print("unminified:", dfa)
dfa = dfa.minify()
print("minified maximum_word_length:", dfa.maximum_word_length())
print("minified:", dfa)
```

**Output**
```shell
unminified maximum_word_length: 4
unminified: DFA(states={'', 'ab', 'abc', 'abcb', 'a'}, input_symbols={'a', 'c', 'b'}, transitions={'': {'a': 'a'}, 'a': {'b': 'ab'}, 'ab': {'c': 'abc'}, 'abc': {'b': 'abcb'}, 'abcb': {}}, initial_state='', final_states={'ab', 'abcb'}, allow_partial=True)
minified maximum_word_length: None
minified: DFA(states={0, 1, 2}, input_symbols={'a', 'c', 'b'}, transitions={0: {'a': 2}, 1: {'c': 2}, 2: {'b': 1}}, initial_state=0, final_states={1}, allow_partial=True)
DFA(states={0, 1, 2, 3}, input_symbols={'a', 'c', '<E>', 'b'}, transitions={0: {'a': 2}, 1: {'c': 2, '<E>': 3}, 2: {'b': 1}, 3: {}}, initial_state=0, final_states={3}, allow_partial=True)
```

**DFA**

**Minified DFA**

**Environment**:

- Python version: 3.9
- automata-lib version: 8.1.0
- Operating System: Debian

The problem seems to be related to the minimization algorithm, potentially in the way it handles transitions and final states during state reduction, but I may have missed something, thanks for your help !

### PR DESCRIPTION
**Pull request by eliotwrobson**
Resolves #182. Modifies the algorithm to deal with partial DFAs correctly. Somewhat inefficient, but the best we can do until Valmari's algorithm gets implemented (which is a separate open issue).

cc @martinec. If you have any more potentially interesting test cases, would be good to add to make sure the algorithm is working correctly (this one is unfortunately a bit tricky to get right).

### CODE REVIEW

File: automata/fa/dfa.py [hunk 1]
```diff
@@ -499,28 +499,47 @@ def _minify(
...
-                for symbol, end_state in path.items():
-                    symbol_dict = transition_back_map[symbol]
-                    # If statement here needed to ignore certain transitions
-                    # when minifying a partial DFA.
-                    if end_state in symbol_dict:
-                        symbol_dict[end_state].append(start_state)
+                for symbol in input_symbols:
+                    end_state = path.get(symbol)
+                    if end_state is not None:
+                        # for symbol, end_state in path.items():
```
Comment (EduardoGoulart1) [inline automata/fa/dfa.py:517]: [minor] remove this comment

File: tests/test_dfa.py [hunk 2]
```diff
@@ -1280,6 +1280,35 @@ def test_minify_partial_dfa(self) -> None:
         self.assertEqual(len(minified_partial_dfa.states), 4)
         self.assertEqual(minified_partial_dfa, partial_dfa_extra_state)
 
+    def test_minify_partial_dfa_correctness(self) -> None:
+        """Test correctness of minifying partial DFAs"""
```
Comment (EduardoGoulart1) [inline tests/test_dfa.py:1284]: For further context, I would add a link to the GitHub issue

**Review by EduardoGoulart1 — APPROVED** [review]
I am currently also struggling with this bug. This LGTM as a temporary fix.

**Comment by eliotwrobson** [comment]
@caleb531 since this is a bug people are actively running into and the volume of changes (outside of the test cases) is small, I'm going to try and merge this sometime today.

File: automata/fa/dfa.py (hunk 1 above)
Comment (eliotwrobson) [reply automata/fa/dfa.py]: Removed here: https://github.com/caleb531/automata/pull/184/commits/ecd87c4e0350cf01e2d721ba5c61b7e9cd2a39f1

File: tests/test_dfa.py (hunk 2 above)
Comment (eliotwrobson) [reply tests/test_dfa.py]: Added comment here: https://github.com/caleb531/automata/pull/184/commits/ecd87c4e0350cf01e2d721ba5c61b7e9cd2a39f1

**Review by caleb531 — APPROVED** [review]

**Comment by caleb531** [comment]
@eliotwrobson Sure, I have no problem with that—the PR looks fine to me. And if you want to go ahead and merge to `main as well, please feel free. Just please be sure to run through all steps in the [Release Checklist](https://github.com/caleb531/automata/blob/main/.github/RELEASE_CHECKLIST.md) (including the creation of the GitHub Release with recent features like the PDA `show_diagram` method, etc.). I just want to make sure that we have everything captured.

**Comment by eliotwrobson** [comment]
@caleb531 going to go ahead and merge since you approved.

I don't quite have time to run through the whole release checklist right now, since we added a couple of new features in addition to this bug fix. I'll try and find time in the next couple of days to push out a minor version. I think the main thing is that people who run into this bug have a fixed version available, and just having that on the develop branch should at least help some.

**Comment by caleb531** [comment]
@eliotwrobson Oh okay, sure. If you think having this fix on `develop` is sufficient, then I'd probably not rather rush a release if we truly don't need to.

**Pull request merged**